### ISSUE TITLE
Possible incorrect data model class for specwcs

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#6361]
_Issue [JP-2289](https://jira.stsci.edu/browse/JP-2289) was created on JIRA by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw):_

In checking new reference file deliveries for MIRI CDP-8 I've noticed a new warning when creating ASDF files from the FITS files delivered:

2021-09-21 15:50:33,079 - stpipe - WARNING - /Users/dlaw/anaconda3/envs/jwst132/lib/python3.9/site-packages/jwst/datamodels/wcs_ref_models.py:61: ValidationWarning: Traceback (most recent call last):
 File "/Users/dlaw/anaconda3/envs/jwst132/lib/python3.9/site-packages/jwst/datamodels/wcs_ref_models.py", line 55, in validate
 assert isinstance(self.model, Model)
AssertionError

It looks like this is occurring when creating old reference files from previous input as well using <https://github.com/STScI-MIRI/miricoord/blob/master/miricoord/mrs/makecrds/makecrds_mrs_cdp8b.py>

The warning itself isn't all that informative, but from digging into the error a bit it looks like it's coming from when I save the reference file, which automatically calls the validation step in <https://github.com/spacetelescope/jwst/blob/master/jwst/datamodels/wcs_ref_models.py.>  The 'distortion' reference file works ok, it goes through the validate defined for

class DistortionMRSModel(ReferenceFileModel)

but the 'specwcs' reference file does not.  It doesn't go through the validate step defined for

class SpecwcsModel(_SimpleModel)

like I'd expect, but instead the step defined for

class _SimpleModel(ReferenceFileModel)

Is there a reason that the SpecwcsModel class is defined as type SimpleModel and not ReferenceFileModel like all of the other classes in wcs_ref_models.py?

If I hack wcs_ref_models to be class SpecwcsModel(_SimpleModel) I get a different error, this time

2021-09-22 11:53:14,025 - stpipe - WARNING - /Users/dlaw/jwcode/jwst/jwst/datamodels/reference.py:47: ValidationWarning: Model.meta is missing values for ['reftype']
 warnings.warn(message, ValidationWarning)

I.e., the reference file doesn't have the reftype keyword set.  I think I see why; the distortion files in wcs_ref_models.py have an explicit

def on_save(self, path=None):
 self.meta.reftype = self.reftype

but the specwcs files do not.  Adding it fixes this problem.

------

So, to summarize, I think class SpecwcsModel should be of type ReferenceFileModel, and that the on_save command should be added.

Running the rest of the way through the code, there's also a wealth of validation errors on the cross-dichroic files not matching BAND strings, which makes me think that many of the validation checks in wcs_ref_models.py are outdated and incorrect.

If [Jonathan Eisenhamer](https://jira.stsci.edu/secure/ViewProfile.jspa?name=eisenham) agrees with the class type change, I'll try to put in a PR myself that fixes all of the errors that I'm seeing pop up.  Discussing line by line in a PR is probably more efficient.

**Comment by stscijgbot-jp** [comment]
_Comment by [James Davies](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies) on [JIRA](https://jira.stsci.edu/browse/JP-2289?focusedCommentId=550497#comment-550497):_

bq. Is there a reason that the SpecwcsModel class is defined as type SimpleModel and not ReferenceFileModel like all of the other classes in wcs_ref_models.py?

Yes, the class inheritance goes

```
ReferenceFileModel > _SimpleModel > SpecwcsModel
```

So that `_SimpleModel` encapsulates everything in `ReferenceFileModel` plus adding some further validation and unit and metadata setting on open and save.

Btw, the `super().validate()` calls the `validate()` method of the superclass before running the rest of `self.validate()`.  So validation for `SpecwcsModel` happens in the following order:

```
ReferenceFileModel.validate()
_SimpleModel.validate()
SpecwcsModel.validate()
```

It sounds like it is failing on that 2nd step?

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-2289?focusedCommentId=550506#comment-550506):_

Yes, it's that second one _SimpleModel.validate() that's failing when run on (e.g.) <https://jwst-crds.stsci.edu/browse/jwst_miri_specwcs_0024.asdf>

Just adding the

def on_save(self, path=None):
self.meta.reftype = self.reftype

to the SpecwcsModel class didn't fix it.  Why is SpecwcsModel different from RegionsModel, WavelengthrangeModel, FilteroffsetModel, etc?

**Comment by stscijgbot-jp** [comment]
_Comment by [Jonathan Eisenhamer](https://jira.stsci.edu/secure/ViewProfile.jspa?name=eisenham) on [JIRA](https://jira.stsci.edu/browse/JP-2289?focusedCommentId=550514#comment-550514):_

So, technically, `_SimpleModel` is a `ReferenceFileModel`, so there should be no particular reason to change the class inheritance.

Reclassing to `ReferenceFileModel` is simply skipping the `validation` method that was causing the original issue.

Now, the question of whether one should be re-classing is an executive decision. I am guessing that `SpecwcsModel` does not have a single, basic Astropy Model for the WCS? The WCS model is more complicated than that? If that is case, then the changes, as specified in the PR, are nearly correct. The only piece missing is the validation of whatever the WCS model is, because that is being bypassed.

For comparison, the WCS model information is more complicated for `DistortionMRSModel`.

The curious question is, in the error situation, what is `self.model` if it is not an Astropy Model?  Where exactly is the error occurring in the aforementioned script, makecrds_mrs_cdp8b? I am asking because if `SpecwcsModel` conceptually does have a simple, single Astropy Model, that should be satisfied by `_SimpleModel`, then why does the validation fail when it should be succeeding.? Is it something in the way the references are being re-written, etc.?

**Comment by stscijgbot-jp** [comment]
_Comment by [James Davies](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies) on [JIRA](https://jira.stsci.edu/browse/JP-2289?focusedCommentId=550515#comment-550515):_

Yeah, the problem is that this datamodel has a list of compound models assigned to `self.model`, one for each slice, but `_SimpleModel` assumes the `self.model` attribute itself is an astropy model, not a list of models.

I suspect this never worked?

A workaround would be to change

<https://github.com/spacetelescope/jwst/blob/c06bb96b3237bb30b0472b6a024ac1769bd2f064/jwst/datamodels/wcs_ref_models.py#L55>

to something like

```
assert isinstance(self.model, Model) or all([isinstance(m, Model) for m in self.model])
```

Would fix it.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-2289?focusedCommentId=550988#comment-550988):_

Fixed by #6362

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-2289?focusedCommentId=550487#comment-550487):_

Filed PR #6362 to make the requested changes more obvious.

**Comment by stscijgbot-jp** [comment]
_Comment by [James Davies [X]](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies) on [JIRA](https://jira.stsci.edu/browse/JP-2289?focusedCommentId=550497#comment-550497):_

bq. Is there a reason that the SpecwcsModel class is defined as type SimpleModel and not ReferenceFileModel like all of the other classes in wcs_ref_models.py?

Yes, the class inheritance goes

```
ReferenceFileModel > _SimpleModel > SpecwcsModel
```

So that `_SimpleModel` encapsulates everything in `ReferenceFileModel` plus adding some further validation and unit and metadata setting on open and save.

Btw, the `super().validate()` calls the `validate()` method of the superclass before running the rest of `self.validate()`.  So validation for `SpecwcsModel` happens in the following order:

```
ReferenceFileModel.validate()
_SimpleModel.validate()
SpecwcsModel.validate()
```

It sounds like it is failing on that 2nd step?

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-2289?focusedCommentId=550506#comment-550506):_

Yes, it's that second one _SimpleModel.validate() that's failing when run on (e.g.) <https://jwst-crds.stsci.edu/browse/jwst_miri_specwcs_0024.asdf>

Just adding the

def on_save(self, path=None):
self.meta.reftype = self.reftype

to the SpecwcsModel class didn't fix it.  Why is SpecwcsModel different from RegionsModel, WavelengthrangeModel, FilteroffsetModel, etc?

**Comment by stscijgbot-jp** [comment]
_Comment by [James Davies [X]](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies) on [JIRA](https://jira.stsci.edu/browse/JP-2289?focusedCommentId=550515#comment-550515):_

Yeah, the problem is that this datamodel has a list of compound models assigned to `self.model`, one for each slice, but `_SimpleModel` assumes the `self.model` attribute itself is an astropy model, not a list of models.

I suspect this never worked?

A workaround would be to change

<https://github.com/spacetelescope/jwst/blob/c06bb96b3237bb30b0472b6a024ac1769bd2f064/jwst/datamodels/wcs_ref_models.py#L55>

to something like

```
assert isinstance(self.model, Model) or all([isinstance(m, Model) for m in self.model])
```

Would fix it.

**Comment by stscijgbot-jp** [comment]
_Comment by [James Davies [X]](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies) on [JIRA](https://jira.stsci.edu/browse/JP-2289?focusedCommentId=550522#comment-550522):_

I'll add that the `assign_wcs` code that uses this "specwcs" reftype reference file for MIRI MRS expects the `model` attribute to be a list of astropy models, one for each slice. The schema for it defines it as such

<https://github.com/spacetelescope/jwst/blob/c06bb96b3237bb30b0472b6a024ac1769bd2f064/jwst/datamodels/schemas/specwcs.schema.yaml#L20-L27>

So clearly the reffile as currently delivered works. I think the reffile(s) just never got properly validated before delivery.  Doing so would have brought up the problem in our validator - the assumption that the model can't be a list of models.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-2289?focusedCommentId=550525#comment-550525):_

Sounds like this is another case of more-aggressive validation recently complaining about files we've been working with for a while.

The suggested fix above seems to work, so I've updated the PR with it and reverted to

class SpecwcsModel(_SimpleModel)

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-2289?focusedCommentId=550988#comment-550988):_

Fixed by #6362
